platform/api/esx/api.go

```diff
@@ -0,0 +1,759 @@
+		return nil, fmt.Errorf("waiting for net ip: %v", err)
+	}
+
+	var pub string
+	var priv string
+
+	for _, ips := range ip {
+		for _, val := range ips {
+			addr := net.ParseIP(val)
+			if addr.IsLinkLocalMulticast() || addr.IsLinkLocalUnicast() {
+				continue
+			}
+			pub = val
+			priv = val
+		}
+	}
+
+	var mvm mo.VirtualMachine
+	err = vm.Properties(ctx, vm.Reference(), []string{"summary", "guest.ipAddress"}, &mvm)
+	if err != nil {
+		return nil, fmt.Errorf("getting machine reference: %v", err)
+	}
+
+	return &ESXMachine{
+		Name:      mvm.Summary.Config.Name,
+		PublicIP:  pub,
+		PrivateIP: priv,
+	}, nil
+}
+
+func New(opts *Options) (*API, error) {
+	if opts.Server == "" || opts.User == "" || opts.Password == "" {
+		profiles, err := auth.ReadESXConfig(opts.ConfigPath)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't read ESX config: %v", err)
+		}
+
+		if opts.Profile == "" {
+			opts.Profile = "default"
+		}
+		profile, ok := profiles[opts.Profile]
+		if !ok {
+			return nil, fmt.Errorf("no such profile %q", opts.Profile)
+		}
+		if opts.Server == "" {
+			opts.Server = profile.Server
+		}
+		if opts.User == "" {
+			opts.User = profile.User
+		}
+		if opts.Password == "" {
+			opts.Password = profile.Password
+		}
+		if opts.BaseVMName == "" {
+			opts.BaseVMName = profile.BaseVMName
+		}
+	}
+
+	esxUrl := fmt.Sprintf("%s:%s@%s", opts.User, opts.Password, opts.Server)
+	u, err := soap.ParseURL(esxUrl)
+	if err != nil {
+		return nil, fmt.Errorf("parsing ESX URL: %v", err)
+	}
+
+	ctx := context.Background()
+
+	client, err := govmomi.NewClient(ctx, u, true)
+	if err != nil {
+		return nil, fmt.Errorf("connecting to ESX: %v", err)
+	}
+
+	return &API{
+		options: opts,
+		client: client,
+		ctx:    ctx,
+	}, nil
+}
+
+func getNetworkDevice(net object.NetworkReference) (types.BaseVirtualDevice, error) {
+	backing, err := net.EthernetCardBackingInfo(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+
+	device, err := object.EthernetCardTypes().CreateEthernetCard("e1000", backing)
+	if err != nil {
+		return nil, err
+	}
+
+	return device, nil
+}
+
+func (a *API) buildCloneSpec(baseVM *object.VirtualMachine, folder *object.Folder, network object.NetworkReference, resourcePool *object.ResourcePool, datastore *object.Datastore, ignitionData string) (*types.VirtualMachineCloneSpec, error) {
+	devices, err := baseVM.Device(a.ctx)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get base VM devices: %v", err)
+	}
+	var card *types.VirtualEthernetCard
+	for _, device := range devices {
+		if c, ok := device.(types.BaseVirtualEthernetCard); ok {
+			card = c.GetVirtualEthernetCard()
+			break
+		}
+	}
+	if card == nil {
+		return nil, fmt.Errorf("No network device found.")
+	}
+
+	netDev, err := getNetworkDevice(network)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get new network backing device: %v", err)
+	}
+
+	card.Backing = netDev.(types.BaseVirtualEthernetCard).GetVirtualEthernetCard().Backing
```

Comment:
I haven't seen it fail and this line of code is based on the govc code for building the cloning spec:

https://github.com/vmware/govmomi/blob/master/govc/vm/clone.go#L289